shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXConnection.scala

```diff
@@ -40,21 +45,98 @@ private[ucx] abstract class UCXTagCallback {
+ */
+private[ucx] abstract class UCXAmCallback {
+  def onError(am: UCXActiveMessage, ucsStatus: Int, errorMsg: String): Unit
+  def onMessageStarted(receiveAm: UcpRequest): Unit
+  def onSuccess(am: UCXActiveMessage, buff: RefCountedDirectByteBuffer): Unit
+  def onCancel(am: UCXActiveMessage): Unit
+
+  // hook to allocate memory on the host
+  // a similar hook will be needed for GPU memory
+  def onHostMessageReceived(size: Long): RefCountedDirectByteBuffer
+}
+
+class UCXServerConnection(ucx: UCX, transport: UCXShuffleTransport)
+  extends UCXConnection(ucx) with ServerConnection {
   override def startManagementPort(host: String): Int = {
     ucx.startManagementPort(host)
   }
 
-  override def send(sendPeerExecutorId: Long, bounceBuffers: Seq[AddressLengthTag],
+  override def send(sendPeerExecutorId: Long, buffer: AddressLengthTag,
     cb: TransactionCallback): Transaction =
-    send(sendPeerExecutorId, null, bounceBuffers, cb)
+    send(sendPeerExecutorId, buffer, Seq.empty, cb)
+
+  override def registerRequestHandler(
+      requestType: RequestType.Value, cb: TransactionCallback): Unit = {
+
+    class RequestAmCallback extends UCXAmCallback {
+      val tx = createTransaction
+      tx.start(UCXTransactionType.Request, 1, cb)
+      override def onSuccess(am: UCXActiveMessage, buff: RefCountedDirectByteBuffer): Unit = {
+        logDebug(s"At requestHandler for ${requestType} and am: " +
+          s"$am")
+        tx.completeWithSuccess(requestType, am.header, Option(buff))
+      }
 
-  override def send(sendPeerExecutorId: Long, header: AddressLengthTag,
-    cb: TransactionCallback): Transaction =
-    send(sendPeerExecutorId, header, Seq.empty, cb)
+      override def onHostMessageReceived(size: Long): RefCountedDirectByteBuffer = {
+        transport.getDirectByteBuffer(size)
+      }
+
+      override def onError(am: UCXActiveMessage, ucsStatus: Int, errorMsg: String): Unit = {
+        tx.completeWithError(errorMsg)
+      }
+
+      override def onCancel(am: UCXActiveMessage): Unit = {
+        tx.completeCancelled(requestType, am.header)
+      }
+
+      override def onMessageStarted(receiveAm: UcpRequest): Unit = {
+        tx.registerPendingMessage(receiveAm)
+      }
+    }
+
+    ucx.registerRequestHandler(
+      UCXActiveMessage(composeRequestAmId(requestType),
+        singleShotCallbackGen = Option(() => new RequestAmCallback)))
+  }
+
+  override def respond(peerExecutorId: Long,
+                       messageType: RequestType.Value,
+                       header: Long,
+                       response: ByteBuffer,
+                       cb: TransactionCallback): Transaction = {
+
+    val tx = createTransaction
+    tx.start(UCXTransactionType.Request, 1, cb)
+
+    logDebug(s"Responding to ${peerExecutorId} at ${TransportUtils.toHex(header)} " +
+      s"with ${response}")
+
+    val responseAm = UCXActiveMessage(composeResponseAmId(messageType), Option(header))
```

Comment:
I picked `Option` because @gerashegalov has beating `Some` out of me, this being the main point:
```
scala> Option(null)
res2: Option[Null] = None

scala> Some(null)
res3: Some[Null] = Some(null)
```
What do you think @jlowe if we keep `Option`?

Reply:
In all the cases I marked, it is known not to be `null`.  It's not must-fix from my perspective, just an unnecessary use of `Option`.